src/coreclr/jit/objectalloc.cpp

```diff
@@ -698,11 +698,24 @@ void ObjectAllocator::MarkEscapingVarsAndBuildConnGraph()
         {
             JITDUMP("   V%02u is address exposed\n", lclNum);
             MarkLclVarAsEscaping(lclNum);
+            continue;
         }
-        else if (lclNum == comp->info.compRetBuffArg)
+
+        if (lclNum == comp->info.compRetBuffArg)
         {
             JITDUMP("   V%02u is retbuff\n", lclNum);
             MarkLclVarAsEscaping(lclNum);
+            continue;
+        }
+
+        // We have to mark all struct params as escaping, because
+        // their GC reporting is controlled by the caller
+        //
+        if (lclDsc->lvIsParam && (lclDsc->lvType == TYP_STRUCT))
+        {
+            JITDUMP("   V%02u is a struct param\n", lclNum);
+            MarkLclVarAsEscaping(lclNum);
+            continue;
         }
```

Comment:
Is this only necessary for implicit byrefs, or truly all struct parameters?

Reply:
Can similar issues happen for primitives parameters too then?

Reply:
Indeed, seems it would be better to create new locals instead of doing this retyping. Probably even for all locals, not just parameter ones.

Reply:
Seems like the caller is responsible for implicit byrefs, otherwise callee. See `GCInfo::gcMakeRegPtrTable`.

If had some way of forcing promotion we could probably lift this limitation too. But for now will just go with blocking implicit byrefs.

Reply:
Implicit byrefs are retyped to pointers later by `fgRetypeImplicitByRefArgs` -- retyping their local here is not going to have any effect on that, so I guess in that sense it makes sense that there would be issues around this.
I wonder if there are any similar issues around OSR locals... I suppose not because this retyping does not happen.